src/verso/Verso/Doc/Elab/Incremental.lean

```diff
@@ -0,0 +1,209 @@
+
+ * `σ` is the DSL elaborator's own monadic state. This may differ from the state used in most of the
+   DSL elaborator's API, because it must capture the _full_ state of `m`. In particular, if `m` is a
+   transformer of `TermElabM`, then `σ` must include `TermElabM`'s state along with the DSL
+   elaboration state.
+
+ * `m` is the DSL's elaboration monad. It must support many of Lean's own elaboration effects, so it
+   will frequently be a transformed `TermElabM`. If so, its `MonadStateOf σ` instance must capture
+   `TermElabM`'s state, e.g. via `saveState` and `TermState.restoreFull`.
+
+ * `cmd` is the complete syntax of the command being elaborated. This is used only for displaying
+   interactive feedback to the user, and it should be a piece of syntax whose range encompasses all
+   the individual steps to be elaborated.
+
+ * `steps` is the individual steps to incrementally elaborate. These should all be contained within
+   `cmd`. They are saved in incremental snapshots and compared (with `Syntax.structRangeEq`) to
+   determine whether the snapshot should be re-used or invalidated.
+
+ * `initAct` contains any monadic actions to run to establish an initial elaboration state prior to
+   elaborating the command steps.
+
+ * `endAct` contains any monadic actions to run after incremental elaboration is completed.
+
+ * `handleStep` describes the DSL's elaboration procedure for a single step from `steps`.
+
+ * `lift` is a means of running the DSL's monad inside `CommandElabM`.
+
+The caller is responsible for resolving the returned promise to indicate to Lean that the command is
+fully elaborated. This should be the last thing it does after any further state changes.
+
+Ordinary exceptions thrown by `handleStep` are caught and logged incrementally, unless they
+represent unrecoverable internal Lean errors, in which case incremental elaboration is terminated.
+-/
+def incrementallyElabCommand
+    [IncrementalSnapshot snapshot σ]
+    [TypeName snapshot] [ToSnapshotTree snapshot] [Nonempty snapshot]
+    [Nonempty σ]
+    [Monad m] [MonadLiftT BaseIO m] [MonadExcept Exception m] [MonadLog m] [AddMessageContext m] [MonadFinally m]
+    [MonadStateOf σ m]
+    (cmd : Syntax)
+    (steps : Array Syntax)
+    (initAct : m Unit)
+    (endAct : m Unit)
+    (handleStep : Syntax → m Unit)
+    (lift : {α : _} → m α → CommandElabM α)
+    : CommandElabM (IO.Promise snapshot × σ) := do
+  if let some snap := (← read).snap? then
+    withReader (fun ρ => {ρ with snap? := none}) do
+      lift <| do
+        let mut oldSnap? := snap.old?.bind (·.val.get.toTyped? (α := snapshot))
+        let mut nextPromise ← IO.Promise.new
+        let initData := {diagnostics := .empty}
+        snap.new.resolve <| DynamicSnapshot.ofTyped <|
+          mkSnap cmd initData (.pure (← get)) (some {range? := cmd.getRange?, task := nextPromise.result})
+        initAct
+        let mut errors : MessageLog := .empty
```

Comment:
I think we should be able to use `m`'s message log instead of keeping track of it manually. I'm not even sure the `finally` does the right thing for local mutable variables.

Reply:
OK - this has been fixed in the latest version. Thanks!